### DISCUSSION TITLE
RFC: Approach to support FIPS mode

### DISCUSSION THREAD

**Discussion by juliedavila** [architect/architect#1246] [Feature Requests]
Under the requirements of FedRAMP, a compliance body which governs security standards for *aaS products there is an explicit requirement that all comms in-transit utilize FIPS certified endpoints. AWS already provides these endpoints for [US East/West](https://aws.amazon.com/compliance/fips/). And leveraging them isn't terribly complex:

```javascript
const DB = require('aws-sdk/clients/dynamodb')
const opts = {
  region: 'us-east-1',
  endpoint: 'dynamodb-fips.us-east-1.amazonaws.com'
}

const db = new DB(opts)

//...do stuff with db objection
```

That being said, I'd like to get consensus on the approach I'd like to take.

## Impact Areas

This implementation would impact Architect in places where there is Lambda => SomeAWSService communication. As far as I can tell, this only exists in  [@architect/functions](https://github.com/architect/functions).

There is nothing to do with regards to deployment logic or anything else.

## General Feature Restriction

This cannot work for ALL deployment models. AWS FIPS endpoints only exist in 

1. us-east-1
2. us-east-2
3. us-west-1
4. us-west-2
5. ca-central-1
6. govcloud

Therefore, adding support for this would also mean having some logic `if FIPS and REGION NOT-IN-FIPS-REGIONS; then ERR` This of course would be specified in docs.

## Approach

We can take a couple of approaches. One driven by `.arc` or one driven by an environment variable (or both with preference given to the env variable).

within `.arc`

```
@aws
fipsMode true
```

and `ARC_FIPS_MODE=1` for the environment variable

## Validation example

From https://github.com/architect/functions/blob/master/src/tables/db.js

We change it to be something like (note, I'd certainly break out the obvious bits to `index.js` or similar where all the other functions could access the same array/validator).

```javascript
let aws = require('aws-sdk')
let https = require('https')
let DB = aws.DynamoDB
let endpoint = new aws.Endpoint('http://localhost:5000')

let fipsRegions = [
  'us-east-1',
  'us-east-2',
  'us-west-1',
  'us-west-2',
  'ca-central-1'
]

if (process.env.ARC_FIPS === '1') {
  if (!fipsRegions.includes(DB.region)) {
    throw new Error(`FIPS mode specified for non FIPS region "${DB.region}". Valid FIPS Regions: ${fipsRegions.join(', ')}`)
  }
  DB.endpoint = DB.endpoint.replace(/(https:\/\/)(\w+)(.+$)/, "$1$2-fips$3")
}

if (typeof process.env.NODE_ENV === 'undefined')
  process.env.NODE_ENV = 'testing'

let testing = process.env.NODE_ENV === 'testing'

if (!testing) {
  let agent = new https.Agent({
    keepAlive: true,
    maxSockets: 50,
    rejectUnauthorized: true,
  })
  aws.config.update({
    httpOptions: { agent }
  })
}

module.exports = testing ? new DB({ endpoint }) : new DB
```

Thoughts?

**Comment by brianleroux** [comment]
I like it; could start w the opt-in through `ARC_FIPS` in `@architect/functions` before ratifying into the `.arc` and `.arc-config` files just to get it goin too.
